staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go

```diff
@@ -284,10 +285,28 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 	return &authenticator.Response{
 		User: &user.DefaultInfo{
 			Name:   chain[0].Subject.CommonName,
```

Comment:
IMO, either:

- We should update the documentation on authentication to make it clear that we always use CN as username, and never UID.
- We should support certificates with userid set but no common name, and in that case we popular both UID and User from userid.

Unfortunately, we use Organization and CommonName in a somewhat off-label formulation, but I also think we'll have to support that approach for many years to come.

Reply:
Good points! I'm open to either approach, and I'm willing to defer to other decision makers.

That being said, I think the first option is preferable because it's simpler: CN sets the `user.Name`, and UID sets the `user.UID`, without any conditionals.

The second option is appealing, but I'm not sure there's a need to fallback (ie: setting `user.Name` from the UID if CN is empty). We can always add it as a followup if a use case arises.

Assuming we proceed with the first option, I'm happy to update the docs once this PR is approved. I found these sections where I can clarify the CN and UID expectations:

https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes
> Kubernetes determines the username from the common name field in the 'subject' of the cert (e.g., "/CN=bob"). 

and "X509 client certificates" here:
https://kubernetes.io/docs/reference/access-authn-authz/authentication/#x509-client-certificates